modules/bmc/redfish.rb

```diff
@@ -0,0 +1,210 @@
+      end
+
+      def identifyoff
+        host.Systems.Members[0].patch(payload: { 'IndicatorLED' => 'Off' })
+      end
+
+      def poweroff(soft=false)
+        poweraction( soft ? 'GracefulShutdown' : 'ForceOff' )
+      end
+
+      def powercycle
+        poweraction('ForceRestart')
+      end
+
+      def poweron
+        poweraction('On')
+      end
+
+      def powerstatus
+        host.Systems.Members[0].PowerState.downcase
+      end
+
+      def poweron?
+        powerstatus == 'on'
+      end
+
+      def poweroff?
+        powerstatus == 'off'
+      end
+
+      def bootdevice
+        host.Systems.Members[0].Boot['BootSourceOverrideTarget']
+      end
+
+      def bootdevices
+        # RedFish will tell you which devices can be put in the boot source override
+        # but Foreman seems structurally arranged at present for these values to be
+        # hardcoded. I can think of a few reasons to continue with that pattern. But
+        # if in the future it becomes advantageous to do something more dynamic --
+        # this is how to find out:
+        #
+        #host.Systems.Members.Boot['BootSourceOverrideTarget@Redfish.AllowableValues']
+        #
+        # Dell gives a list like [ None, Floppy, Cd, Hdd, SDCard, Utilities,
+        #                          BiosSetup, Pxe ]
+        # HPE -        [ None, Floppy, Cd, Hdd, Usb, Utilities, BiosSetup, Pxe ]
+        # SuperMicro - [ None, Pxe, Hdd, Diags, Cd, BiosSetup, FloppyRemovableMedia,
+        #                UsbKey, UsbHdd, UsbFloppy, UsbCd, UefiCd, UefiHdd, ... ]
+        #
+        # For the four devices Foreman hardcodes, lucky for us all three use the same set
+        # of values.
+        [ 'pxe', 'disk', 'bios', 'cdrom' ]
+      end
+
+      def bootdevice=(args={ :device => nil, :reboot => false, :persistent => false })
+        devmap = { 'bios'  => 'BiosSetup',
+                   'cdrom' => 'Cd',
+                   'disk'  => 'Hdd',
+                   'pxe'   => 'Pxe' }
+
+        host.Systems.Members[0].patch(
+          payload: {
+            'Boot' => {
+              'BootSourceOverrideTarget' => devmap[args[:device]],
+              'BootSourceOverrideEnabled' => args[:persistent] ? 'Enabled' : 'Once'
+            }
+          })
+        powercycle if reboot
+      end
+
+      def bootpxe(reboot=false, persistent=false)
+        bootdevice = { :device => 'pxe', :reboot => reboot, :persistent => persistent }
+      end
+
+      def bootdisk(reboot=false, persistent=false)
+        bootdevice = { :device => 'disk', :reboot => reboot, :persistent => persistent }
+      end
+
+      def bootbios(reboot=false, persistent=false)
+        bootdevice = { :device => 'bios', :reboot => reboot, :persistent => persistent }
+      end
+
+      def bootcdrom(reboot=false, persistent=false)
+        bootdevice = { :device => 'cdrom', :reboot => reboot, :persistent => persistent }
+      end
```

Comment:
Are these good candidates to be moved to the Base class?

I don't like method names, these are "side effect" methods which should be named with bang e.g. `bootpxe!` but I see almost zero value in these, does not make the code much shorter.

Reply:
Are you asking me that? My hastily formed opinion is, maybe not for the base class, since the ssh and shell methods don't (can't) provide them. I have tried to avoid doing much more than implementing Redfish as if it were effectively another IPMI provider, and trying not to make too many philosophical decisions about how I think existing code "should" work instead. I've deliberately preferred to work as much as possible with what's already present, thinking (perhaps naively) that it would help keep the focus on getting Redfish support added in, and not drift into philosophical debates about coding convention or quality -- my opinion on that has been, you guys are the maintainers, it's your call and not mine. I should be clear, too, I'm not complaining about how anything has gone so far!

Regarding the method names and the value of them, the API code calls these methods. I don't have anything to say if you want to change the way that all works. But would involve a more invasive change.

Reply:
I was just thinking out loud. True, these are not implemented SSH fake implementation.

The other comment was just a nitpicking - it's a common practice in Ruby to end methods with `bang!` characters if they only have a side effect and does not return a relevant value. Feel free to ignore this.